### PR TITLE
Removed bootstrap asserts

### PR DESCRIPTION
**Pull request by GentBinaku**

### CODE REVIEW

File: src/bootstrap/job.rs [hunk 1]
```diff
@@ -54,6 +54,9 @@ pub unsafe fn setup(build: &mut Build) {
 
     // Create a new job object for us to use
     let job = CreateJobObjectW(ptr::null_mut(), ptr::null());
+    if job.is_null() {
+        eprintln!("{}", io::Error::last_os_error());
+    }
     assert!(!job.is_null(), "{}", io::Error::last_os_error());
```
Comment (Nilstrieb) [inline src/bootstrap/job.rs:60]: This seems redundant

**Review by Noratrieb — COMMENTED** [review]
Maybe it would make sense to create a `exit_error!` macro that wraps the `eprintln` and `detail_exit`?

File: src/bootstrap/builder.rs [hunk 2]
```diff
@@ -466,11 +469,12 @@ impl<'a> ShouldRun<'a> {
...
-            self.kind == Kind::Setup || !self.builder.src.join(alias).exists(),
-            "use `builder.path()` for real paths: {}",
-            alias
-        );
+        match self.kind == Kind::Setup || !self.builder.src.join(alias).exists() {
+            true => {}
+            false => {
+                eprintln!("use `builder.path()` for real paths: {:?}", alias)
+            }
+        };
```
Comment (ozkanonur) [inline src/bootstrap/builder.rs:477]: `if` statement would better fit here

File: src/bootstrap/config.rs [hunk 3]
```diff
@@ -1187,7 +1189,13 @@ impl Config {
...
             config.llvm_from_ci = match llvm.download_ci_llvm {
                 Some(StringOrBool::String(s)) => {
-                    assert!(s == "if-available", "unknown option `{}` for download-ci-llvm", s);
+                    match s == "if-available" {
+                        false => {
+                            eprintln!("unknown option `{}` for download-ci-llvm", s);
+                            crate::detail_exit(1);
+                        }
+                        true => {}
+                    };
```
Comment (ozkanonur) [inline src/bootstrap/config.rs:1198]: Same here as well

**Review by onur-ozkan — CHANGES_REQUESTED** [review]
I agree suggestion of @Nilstrieb, abstracting those `eprintln`, `crate::detail_exit` duplications can be better for future changes.

File: src/bootstrap/builder.rs (hunk 2 above)
Comment (jyn514) [reply src/bootstrap/builder.rs:477]: this is not a user-facing error message, this only happens if there's a bug in a Step. so I think an assert like before is better.

**Review by jyn514 — CHANGES_REQUESTED** [review]
I've only glanced over this, but at a first glance it doesn't look right, it's changed *every* assert to an error message. I want to instead distinguish user errors from internal errors. I'm happy to give guidance on the difference but it may be a while before I can go line by line through the PR.

As a general rule:
- If the panic can only happen because bootstrap's code itself was changed, it's an internal error (e.g. the Step checking in builder.rs)
- If the panic can only happen because of a config.toml option or flags passed to bootstrap, before any commands are run, it's a user error
- If the panic happens because of an external command it's ambiguous, but I would prefer to default to panicking so people report the bug.
- If the panic happens because of an I/O error it's likely an internal error.
- If the panic should never happen (e.g. "path traversal attack") it's an internal error

**Comment by jyn514** [comment]
Maybe a simpler rule is "would it be helpful to show a backtrace of where the error is emitted?" If so it should be a panic, if not it should use eprintln.

**Comment by ozkanonur** [comment]
> Maybe a simpler rule is "would it be helpful to show a backtrace of where the error is emitted?" If so it should be a panic, if not it should use eprintln.

IMO having backtrace on errors always nicer than not having it.

**Comment by jyn514** [comment]
I don't agree, "option X can't be combined with download-ci-llvm" is clear on its own and doesn't need a backtrace. And using panics for expected errors makes it more likely that people will ignore real bugs that we do want them to report.

**Comment by ozkanonur** [comment]
> using panics for expected errors makes it more likely that people will ignore real bugs that we do want them to report.

I didn't look from this perspective. Seems reasonable and makes sense.

**Comment by JohnCSimon** [comment]
@GentBinaku 
ping from triage - can you post your status on this PR? 

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by jyn514** [comment]
I'm going to close this since it's been a while - feel free to reopen if you have time to work on it again.

**Pull request closed without merging**